ldmsend/ldmsend.1
.TH LDMSEND 1 "2010-03-10"
.SH NAME
ldmsend - ldm data source example client
.SH "SYNOPSIS"
\fBldmsend \fR[\fB\-vx\fR] [\fB\-l \fIlogdest\fR] 
[\fB-h \fIremote\fR]
[\fB-P \fIport\fR]
[\fB-s \fIseqno\fR]
[\fB-f \fIfeedtype\fR]
\fIfilename\fR ...
.SH DESCRIPTION
.LP
This program is an example of how to send data using ldm protocols.
Its primary function is as a coding example. It may also be used to
test the ldm server.  It might also be useful to resend data
after a downstream node has been offline. For most applications,
the use of this program is deprecated in favor of \fBpqinsert(1)\fR.
.LP
The program sends each file given on the command line as a separate
ldm data product to the ldm server on \fBremote\fR. The file name is
used as product identifier.
The product feed type is specified by the \fB\-f\fR option.
The current time is used as product timestamp.
.SH OPTIONS 
.TP 8
.BI \-f " feedtype"
Asserts that the data being send has feed type \fIfeedtype\fR.
The default is `EXP', indicating an experimental feedtype.
.TP
.BI \-h " remote"
Sends data to the ldm server on \fIremote\fR. The default is `localhost'.
.TP
.BI \-l " logdest"
Log to \fIlogdest\fP. One of \fB''\fP (system logging daemon), \fB'-'\fP
(standard error stream), or file \fIlogdest\fP. Default is the standard error
stream if the process has a controlling terminal (i.e., the process isn't a
daemon); otherwise, either the LDM log file or the system logging daemon
(execute this program with just the option \fB'-?'\fP to determine which).
.TP
.BI \-P " port"
The port on the remote sytem to which to connect.  Only useful for LDM 6
protocols.  The well-known LDM port is 388.
.TP
.BI \-s " seqno"
Sets the product sequence number associated with the first product
to \fIseqno\fR. Subsequent products (files on the command line)
are numbered sequentially. Defaults to zero.
.TP
.B \-v
Verbose flag.  If present, the log will contain a line for 
each product file that was sent.
.TP
.BI \-x
Requests that debugging information also be printed.
.SH EXAMPLE
The command
.RS +4
.nf
ldmsend -v -h otherhost -f MCIDAS 92010100.are
.fi
.RE
will ship the file `92010100.are' to `otherhost' with feedtype `MCIDAS'.
The ldm server on `otherhost'
should be configured to do something with a McIDAS product whose identifier is
`92010100.are'.
.SH "SEE ALSO"
pqinsert(1), ldmd(1), ldmd.conf(5)
.SH DIAGNOSTICS
Error messages and log messages are written to the log file. 


CHANGE_LOG
6.13.0.0	2016-02-11T12:02:01-0700
    Logging:
        Re-wrote LDM logging because:
            1) It was the major impediment to correctly using the LDM;
            2) It was the major impediment to putting the LDM in a Docker
               container; and
            3) Some system logging daemons sometimes dropped LDM log messages.
        For backward-compatibility, the original logging component, "ulog",
        still exists and is unconditionally included in the LDM library,
        "libldm". Indeed, the original behavior of the LDM can be obtained by
        specifying the option "--with-ulog" to the configure(1) script. By
        default, however, the new logging component, named "slog" (for "simple
        logging"), will be built and used. While "slog" can be told to write to
        the system logging daemon, by default it will write directly to the LDM
        log file if the process doesn't have a controlling terminal (i.e., is a
        daemon) or to the standard error stream if the process does have a
        controlling terminal (i.e., is not a daemon). Like the original "ulog"
        component, "slog" also supports log file rotation via a
        crontab(1)-executed "ldmadmin newlog" command. See the installed HTML
        documentation "$HOME/share/doc/ldm/basics/logging-format.html" or the
        web page
        <http://www.unidata.ucar.edu/software/ldm/ldm-current/basics/logfile-format.html>
        for more information.

6.12.15.42	2016-01-12T10:40:24-0700
    Make test for C++11-compatible g++ more robust.

6.12.15.41	2016-01-12T09:28:01-0700
    Get "make distclean" to work on both Winters (with C++11 support) and Spock
    (without C++11 support).

6.12.15.40	2016-01-11T16:07:56-0700
    Removed mylog2ulog stuff from ulog makefile.
    
    Eliminated memory-leaks from components that read GEMPAK tables.

6.12.15.39	2015-11-16T16:04:41-0700
    mldm_sender_manager(3):
        Add "-y" and "-z" options to execvp(2) of mldm_sender(1).

    pqact.c:
        Added SIGHUP forwarding.

6.12.15.38	2015-11-03T13:07:13-0700
    ldmd.c:
        Corrected some Coverity-discovered lint.

6.12.15.37	2015-11-03T09:41:04-0700
    Increase resolution of printed product-creation timestamp from millisecond
    to microsecond.

6.12.15.36	2015-11-02T15:59:18-0700
    ldmadmin(1):
        Reduce overestimation of new queue capacities (i.e., data capacity,
        product capacity) and don't compute new capacities if the data is
        insufficient.
        
    logging:
        Added logging options for using microsecond resolution in timestamps and
        formatting them according to the ISO 8601 standard. Added respective
        options "-y" and "-z" to ldmd(1) and notifyme(1).

6.12.15.35	2015-10-28T15:29:59-0600
    Create mcast_test(1) for testing LDM-7 on GENI.

6.12.15.34	2015-10-26T16:41:04-0600
    PerProdNotifier:
        Fix setting of product-information by notify_of_bop().

6.12.15.33	2015-10-23T15:57:05-0600
    mldm_sender():
        Moved addition of product-index-to-product-signature entry to before the
        sending of the product because the receiving LDM-7 would sometimes
        request a data-product before the mapping was entered resulting in the
        product not being found.

6.12.15.32	2015-10-22T19:32:42-0600
    Improve receiver session-memory API.

6.12.15.31	2015-10-20T17:11:53-0600
    down7.c:
        Correct location of down7_incNumProds().

        Add mutex destruction to down7_free().

    mldm_sender:
        Add opening of product-queue in thread-safe mode to prevent corruption
        of the regions-in-use list by multiple thread access.

    pq.c:
        Add unlocking of mutex in pq_sequence() before calling `ifMatch()`
        to avoid deadlock in calling a foreign function.

        Add mutex destruction to freeing.

6.12.15.30	2015-10-19T16:23:06-0600
    noaaportIngester(1):
        Added "+000 001 008 010 Total precipitation" to GEMPAK tables.

    Multicast:
        Add feature to lock products until VCMTP sender releases them.

        Modify VCMTP port numbers to match WireShark plugin.

        Improve log messages.

    pq(3):
        Make pq_release() thread-safe.

6.12.15.29	2015-10-08T18:03:25-0600
    Add diagnostic print statements and modify up7-down7_test to try
    diagnosing product-queue corruption seen by Shawn chen.

6.12.15.28	2015-10-06T16:03:42-0600
    pq(3):
        Modified fb_get() to log fblk metrics when it fails to help diagnose
        multicast problem.

6.12.15.27	2015-10-06T12:01:45-0600
    pqutil(1):
        Removed minor lint revealed by Coverity Scan.

    pq(3):
        Use correct product size in pqe_insert() and pq_insertNoSig().

    mcast_lib(3):
        Adapt to new checks by pqe_insert().

6.12.15.26	2015-10-06T09:48:58-0600
    pqact(1):
        Correct the setting of the product-queue cursor based on the last
        successfully-processed data-product from the previous session.
        Before this, pqact(1) might miss a product due to a restart of the LDM.
        This was most apparent in processing NEXRAD-2 data-products. Thanks
        Ryan.

    Product-Queue (pq(3)):
        Refactored pqe_insert() for better error-handling and return status.

        Harmonized return code names by adding PQ_DUP and PQ_BIG and making
        PQUEUE_DUP and PQUEUE_BIG aliases.

6.12.15.25	2015-10-01T15:28:56-0600
    Remove ignoring of SIGCONT from mldm_sender_manager because it's not needed.

    Generalize TCP keep-alive in TcpSend.cpp.

6.12.15.24	2015-09-30T15:40:21-0600
    Reduce number of "make distcheck" to just one in CI-commit.sh.

6.12.15.23	2015-09-30T14:33:58-0600
    Replace "$@" with just $@ in deploy.sh.

6.12.15.22	2015-09-30T13:32:25-0600
    Replace ${configOpts} with "$@" in deploy.sh to accommodate argument with
    space.

6.12.15.21	2015-09-29T16:57:17-0600
    Corrected setting of configure(1) options in delivery/deploy.sh.

6.12.15.20	2015-09-29T15:03:45-0600

ldmsend/ldmsend.c
/* 
 * Sends files to an LDM as data-products.
 *
 * See file ../COPYRIGHT for copying and redistribution conditions.
 */

#define TIRPC
#include <config.h>

#include <assert.h>
#include <errno.h>
#include <fcntl.h>
#include <limits.h>
#include <rpc/rpc.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mman.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <time.h>
#include <unistd.h>

#include "exitStatus.h"
#include "ldm.h"        /* needed by following */
#include "LdmProxy.h"
#include "atofeedt.h"
#include "error.h"
#include "globals.h"
#include "remote.h"
#include "inetutil.h"
#include "ldm_clnt_misc.h"
#include "ldmprint.h"
#include "md5.h"
#include "prod_class.h"
#include "rpcutil.h"
#include "log.h"

#ifdef NO_ATEXIT
#include "atexit.h"
#endif

#ifndef DEFAULT_FEEDTYPE
        /* default to using the "experimental" feedtype */
#define DEFAULT_FEEDTYPE EXP
#endif

static const char*      remote = NULL; /* hostname of data remote */
static LdmProxy*        ldmProxy = NULL;

static void
usage(
    char *av0 /*  id string */)
{
    (void)fprintf(stderr,
"Usage: %s [options] filename ...\n\tOptions:\n", av0);
    (void)fprintf(stderr,
"\t-v           Verbose, tell me about each product\n");
    (void)fprintf(stderr,
"\t-l dest      Log to `dest`. One of: \"\" (system logging daemon), \"-\"\n"
"\t             (standard error), or file `dest`. Default is \"%s\"\n",
                log_get_default_destination());
    (void)fprintf(stderr,
"\t-h remote    remote service host, defaults to \"localhost\"\n");
    (void)fprintf(stderr,
"\t-s seqno     set initial product sequence number to \"seqno\", defaults to 0\n");
    (void)fprintf(stderr,
"\t-f feedtype  assert your feed type as \"feedtype\", defaults to \"%s\"\n", s_feedtypet(DEFAULT_FEEDTYPE));
    exit(1);
}

void
cleanup(void)
{
    if (ldmProxy != NULL) {
        lp_free(ldmProxy);
        ldmProxy = NULL;
    }
    (void)log_fini();
}

static void
signal_handler(int sig)
{
#ifdef SVR3SIGNALS
    /* 
     * Some systems reset handler to SIG_DFL upon entry to handler.
     * In that case, we reregister our handler.
     */
    (void) signal(sig, signal_handler);
#endif
    switch(sig) {
      case SIGINT :
         exit(1);
         return;
      case SIGTERM :
         done = 1;
         return;
      case SIGPIPE :
         exit(1);
    }
}

static void
set_sigactions(void)
{
#ifdef HAVE_SIGACTION
    struct sigaction sigact;

    sigact.sa_handler = signal_handler;
    sigemptyset(&sigact.sa_mask);

    sigact.sa_flags = 0;
    (void) sigaction(SIGINT, &sigact, NULL);

    sigact.sa_flags |= SA_RESTART;
    (void) sigaction(SIGTERM, &sigact, NULL);
    (void) sigaction(SIGPIPE, &sigact, NULL);

    sigact.sa_handler = SIG_IGN;
    (void) sigaction(SIGALRM, &sigact, NULL);
#else
    (void) signal(SIGINT, signal_handler);
    (void) signal(SIGTERM, signal_handler);
    (void) signal(SIGPIPE, signal_handler);
    (void) signal(SIGALRM, SIG_IGN);
#endif
}

static int
fd_md5(MD5_CTX *md5ctxp, int fd, off_t st_size, signaturet signature)
{
    ssize_t     nread;
    char        buf[8192];

    MD5Init(md5ctxp);

    for (; exitIfDone(1) && st_size > 0; st_size -= (off_t)nread) {
        nread = read(fd, buf, sizeof(buf));
        if(nread <= 0) {
            log_syserr("fd_md5: read");
            return -1;
        } /* else */
        MD5Update(md5ctxp, (unsigned char *)buf, (unsigned int)nread);
    }

    MD5Final((unsigned char*)signature, md5ctxp);
    return 0;
}

/*
 * Sends a single, open file to an LDM as a data-product. The number of bytes
 * to be sent is specified by the data-product's metadata. The bytes start at
 * the beginning of the file.
 *
 * Arguments:
 *      proxy           The LDM proxy data-structure.
 *      fd              The file-descriptor open on the file to be sent.
 *      info            The data-product's metadata. Must be completely set.
 *
 * Returns:
 *      0                       Success.
 *      SYSTEM_ERROR            O/S failure. "log_add()" called.
 *      CONNECTION_ABORTED      The connection was aborted. "log_add()"
 *                              called.
 */
static int
send_product(
    LdmProxy*           proxy,
    int                 fd,
    prod_info* const    info)
{
    int                 status;
    product             product;

    product.info = *info;
    product.data = mmap(NULL, info->sz, PROT_READ, MAP_PRIVATE, fd, 0);

    if (MAP_FAILED == product.data) {
        log_syserr("Couldn't memory-map file");
        status = SYSTEM_ERROR;
    }
    else {
        status = lp_send(proxy, &product);
        if (LP_UNWANTED == status) {
            log_notice("Unwanted product: %s", s_prod_info(NULL, 0, info,
                        log_is_enabled_debug));
            status = 0;
        }
        (void)munmap(product.data, info->sz);
    }                                           /* file is memory-mapped */

    return status;
}

/*
 * Sends a list of files to the LDM as data-products.
 *
 * Arguments:
 *      ldmProxy        The LDM proxy data-structure.
 *      offer           The description of the class of data-products that this
 *                      process is willing to send.
 *      origin          The identifier of the host that created the
 *                      data-products (typically the host running this program).
 *      seq_start       The starting value of the data-product sequence number.
 *      nfiles          The number of files to send.
 *      filenames       The pathnames of the files to send.
 *
 * Returns:
 *      0                       Success.
 *      SYSTEM_ERROR            O/S failure. "log_add()" called.
 *      CONNECTION_ABORTED      The connection was aborted. "log_add()"        *                              called.
 */
static int
ldmsend(
    LdmProxy*           ldmProxy,
    prod_class_t*       offer,
    char*               origin,
    int                 seq_start,
    int                 nfiles,
    char*               filenames[])
{
    int                 status = 0;
    char*               filename;
    int                 fd;
    struct stat         statb;
    prod_info           info;
    MD5_CTX*            md5ctxp = NULL;
    prod_class_t*       want;

    /*
     * Allocate an MD5 context
     */
    md5ctxp = new_MD5_CTX();
    if (md5ctxp == NULL)
    {
        log_syserr("new_md5_CTX failed");
        return SYSTEM_ERROR;
    }

    status = lp_hiya(ldmProxy, offer, &want);

    if (status != 0) {
        status = CONNECTION_ABORTED;
    }
    else {
        /* These members are constant over the loop. */
        info.origin = origin;
        info.feedtype = offer->psa.psa_val->feedtype;

        for (info.seqno = seq_start; exitIfDone(1) && nfiles > 0;
                filenames++, nfiles--, info.seqno++) {
            filename = *filenames;
            info.ident = filename;
            /*
             * ?? This could be the creation time of the file.
             */
            (void) set_timestamp(&info.arrival);

            /*
             * Checks 'arrival', 'feedtype', and 'ident'
             * against what the other guy has said he wants.
             */
            if (!prodInClass(offer, &info)) {
                log_info("Not going to send %s", filename);
                continue;       
            }
            if (!prodInClass(want, &info)) {
                log_info("%s doesn't want %s", lp_host(ldmProxy), filename);
                continue;       
            }

            fd = open(filename, O_RDONLY, 0);
            if (fd == -1) {
                log_syserr("open: %s", filename);
                continue;
            }

            if (fstat(fd, &statb) == -1) {
                log_syserr("fstat: %s", filename);
                (void) close(fd);
                continue;
            }

            log_info("Sending %s, %d bytes", filename, statb.st_size);
            
            /* These members, and seqno, vary over the loop. */
            if (fd_md5(md5ctxp, fd, statb.st_size, info.signature) != 0) {
                (void) close(fd);
                continue;
            }
            if (lseek(fd, 0, SEEK_SET) == (off_t)-1) {
                log_syserr("rewind: %s", filename);
                (void) close(fd);
                continue;
            }

            info.sz = (u_int)statb.st_size;

            (void)exitIfDone(1);

            status = send_product(ldmProxy, fd, &info);

            (void) close(fd);

            if (0 != status) {
                log_add("Couldn't send file \"%s\" to LDM", filename);
                break;
            }
        }                                       /* file loop */

        if (lp_flush(ldmProxy))
            log_add("Couldn't flush connection");

        free_prod_class(want);
    }                                           /* HIYA succeeded */

    free_MD5_CTX(md5ctxp);  
    return status;
}


/*
 * Returns:
 *      0               Success
 *      SYSTEM_ERROR    O/S failure. "log_add()" called.
 *      LP_TIMEDOUT     The RPC call timed-out. "log_add()" called.
 *      LP_RPC_ERROR    RPC error. "log_add()" called.
 *      LP_LDM_ERROR    LDM error. "log_add()" called.
 */
int
main(
    int         ac,
    char*       av[])
{
    char            myname[_POSIX_HOST_NAME_MAX];
    char*           progname = av[0];
    prod_class_t    clss;
    prod_spec       spec;
    int             seq_start = 0;
    int             status;
    ErrorObj*       error;
    unsigned        remotePort = LDM_PORT;

    /*
     * Set up error logging
     */
    (void)log_init(progname);

    remote = "localhost";

    (void)set_timestamp(&clss.from);
    clss.to = TS_ENDT;
    clss.psa.psa_len = 1;
    clss.psa.psa_val = &spec;
    spec.feedtype = DEFAULT_FEEDTYPE;
    spec.pattern = ".*";

    {
        extern int optind;
        extern char *optarg;
        int ch;

        while ((ch = getopt(ac, av, "vxl:h:f:P:s:")) != EOF)
            switch (ch) {
            case 'v':
                if (!log_is_enabled_info)
                    (void)log_set_level(LOG_LEVEL_INFO);
                break;
            case 'x':
                (void)log_set_level(LOG_LEVEL_DEBUG);
                break;
            case 'l':
                (void)log_set_destination(optarg);
                break;
            case 'h':
                remote = optarg;
                break;
            case 'f':
                spec.feedtype = atofeedtypet(optarg);
                if(spec.feedtype == NONE)
                {
                    fprintf(stderr, "Unknown feedtype \"%s\"\n",
                            optarg);
                        usage(progname);        
                }
                break;
            case 'P': {
                char*       suffix = "";
                long        port;

                errno = 0;
                port = strtol(optarg, &suffix, 0);

                if (0 != errno || 0 != *suffix ||
                    0 >= port || 0xffff < port) {

                    (void)fprintf(stderr, "%s: invalid port %s\n",
                         av[0], optarg);
                    usage(av[0]);   
                }

                remotePort = (unsigned)port;

                break;
            }
            case 's':
                seq_start = atoi(optarg);
                break;
            case '?':
                usage(progname);
                break;
            }

        ac -= optind; av += optind;

        if(ac < 1) usage(progname);
    }

    /*
     * Register the exit handler
     */
    if(atexit(cleanup) != 0)
    {
        log_syserr("atexit");
        exit(SYSTEM_ERROR);
    }

    /*
     * Set up signal handlers
     */
    set_sigactions();

    (void) strncpy(myname, ghostname(), sizeof(myname));
    myname[sizeof(myname)-1] = 0;

    (void)exitIfDone(INTERRUPTED);

    /*
     * Connect to the LDM.
     */
    status = lp_new(remote, &ldmProxy);

    if (0 != status) {
        log_flush_error();
        status = (LP_SYSTEM == status)
            ? SYSTEM_ERROR
            : CONNECTION_ABORTED;
    }
    else {
        log_debug("version %u", lp_version(ldmProxy));

        status = ldmsend(ldmProxy, &clss, myname, seq_start, ac, av);

        if (0 != status)
            log_flush_error();

        lp_free(ldmProxy);
        ldmProxy = NULL;
    }                                       /* "ldmProxy" allocated */

    return status; 
}
